[JSC] Shrink ArrayAllocationProfile from 16 to 8 bytes

tadeuzagallo opened the pull request:
#### 15ea71cc4b7df927a893c0439ab5ca30a852edbb
<pre>
[JSC] Shrink ArrayAllocationProfile from 16 to 8 bytes
<a href="https://bugs.webkit.org/show_bug.cgi?id=252056">https://bugs.webkit.org/show_bug.cgi?id=252056</a>
rdar://105277033

Reviewed by Yusuke Suzuki.

Currently, `ArrayAllocationProfile`s use 16 bytes: `IndexingMode` (1 byte), the
largest vector length seen (4 bytes), and a pointer to the last seen `JSArray`.
However, the vector length is limited to `BASE_CONTIGUOUS_VECTOR_LEN_MAX`, which
is `25`, so we can easily fit it in a byte, and combined with `CompactPointerTuple`,
we can fit the whole thing in 8 bytes.

* Source/JavaScriptCore/bytecode/ArrayAllocationProfile.cpp:
(JSC::ArrayAllocationProfile::updateProfile):
* Source/JavaScriptCore/bytecode/ArrayAllocationProfile.h:
(JSC::ArrayAllocationProfile::selectIndexingType):
(JSC::ArrayAllocationProfile::vectorLengthHint):
(JSC::ArrayAllocationProfile::updateLastAllocation):
(JSC::ArrayAllocationProfile::updateLastAllocationFor):
(JSC::ArrayAllocationProfile::IndexingTypeAndVectorLength::IndexingTypeAndVectorLength):
(JSC::ArrayAllocationProfile::IndexingTypeAndVectorLength::indexingType const):
(JSC::ArrayAllocationProfile::IndexingTypeAndVectorLength::vectorLength const):
(JSC::ArrayAllocationProfile::IndexingTypeAndVectorLength::operator bool const):

Canonical link: <a href="https://commits.webkit.org/260192@main">https://commits.webkit.org/260192@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4c199e089a510a21d1feb8a752ec83201ee113bb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107486 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16548 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40389 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116649 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116070 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17973 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7790 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99650 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113247 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13575 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96756 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/41202 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95480 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28382 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96823 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9554 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29735 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/96245 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/7553 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10211 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/6633 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/30809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15710 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49316 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/105093 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7049 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11771 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26062 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/bytecode/ArrayAllocationProfile.h:
```diff
@@ -62,17 +62,24 @@ class ArrayAllocationProfile {
     {
-        m_lastArray = lastArray;
+        ASSERT(!isCompilationThread());
+        m_lastArray = IndexingTypeAndVectorLength(lastArray->indexingType(), lastArray->getVectorLength());
```
The reason why we keep a pointer, and not getting the indexing type & vector length at this point is that, they can be changed later, and we would like to capture the final results.
For example, ArrayAllocation site would see VectorLength = 0 and Undecided shape initially. And while this array is used in the user's program, it will get decided shape (e.g. Int32Shape), and it gets the vector length that is suitable for the user's program. And by putting a pointer here first and reading this data while compiling, we would like to capture this "final shape of the allocated array". So, I think this change is not successfully capturing the final shape of arrays allocated from this allocation site.

tadeuzagallo replied:
That's a great point, thanks!

Constellation commented on Source/JavaScriptCore/bytecode/ArrayAllocationProfile.cpp:
```diff
@@ -53,19 +53,20 @@ void ArrayAllocationProfile::updateProfile()
-        IndexingType indexingType = leastUpperBoundOfIndexingTypes(m_currentIndexingType & IndexingTypeMask, lastArray->indexingType());
-        if (isCopyOnWrite(m_currentIndexingType)) {
+        IndexingType indexingType = leastUpperBoundOfIndexingTypes(current().indexingType() & IndexingTypeMask, lastArray->indexingType());
+        if (isCopyOnWrite(current().indexingType())) {
```
Probably I think we should read `storage`'s indexingType since `current` can be concurrently updated now.

tadeuzagallo replied:
I thought about it, but for some reason I thought it wasn't a problem... we only call `setPointer` concurrently, and since it's only a word, it should be safe, no?  In any case, I can update it.

Constellation commented on Source/JavaScriptCore/bytecode/ArrayAllocationProfile.cpp:
```diff
@@ -53,19 +53,20 @@ void ArrayAllocationProfile::updateProfile()
-        m_currentIndexingType = indexingType;
-        m_largestSeenVectorLength = std::min(std::max(m_largestSeenVectorLength, lastArray->getVectorLength()), BASE_CONTIGUOUS_VECTOR_LEN_MAX);
+        unsigned largestSeenVectorLength = std::min(std::max(current().vectorLength(), lastArray->getVectorLength()), BASE_CONTIGUOUS_VECTOR_LEN_MAX);
+        m_storage.setType(IndexingTypeAndVectorLength(indexingType, largestSeenVectorLength));
```
Let's put `static_assert(BASE_CONTIGUOUS_VECTOR_LEN_MAX < UINT8_MAX)` to ensure that vector-length is within uint8_t.

tadeuzagallo replied:
The assertion already exists in `IndexingTypeAndVectorLength`

Constellation commented on Source/JavaScriptCore/bytecode/ArrayAllocationProfile.cpp:
```diff
@@ -53,19 +53,20 @@ void ArrayAllocationProfile::updateProfile()
         }
-        m_currentIndexingType = indexingType;
-        m_largestSeenVectorLength = std::min(std::max(m_largestSeenVectorLength, lastArray->getVectorLength()), BASE_CONTIGUOUS_VECTOR_LEN_MAX);
+        unsigned largestSeenVectorLength = std::min(std::max(current().vectorLength(), lastArray->getVectorLength()), BASE_CONTIGUOUS_VECTOR_LEN_MAX);
```
Ditto for `current`.

Constellation commented on Source/JavaScriptCore/bytecode/ArrayAllocationProfile.h:
```diff
@@ -85,18 +90,60 @@ class ArrayAllocationProfile {
+
+        // The format is: (IndexingType << 8) | VectorLength
+        static constexpr uint16_t s_vectorLengthMask = static_cast<uint8_t>(-1);
+        static constexpr uint16_t s_indexingTypeShift = 8;
```
`s_` is not necessary.

Constellation reviewed (approved):
r=me

tadeuzagallo committed:
[JSC] Shrink ArrayAllocationProfile from 16 to 8 bytes
https://bugs.webkit.org/show_bug.cgi?id=252056
rdar://105277033

Reviewed by Yusuke Suzuki.

Currently, `ArrayAllocationProfile`s use 16 bytes: `IndexingMode` (1 byte), the
largest vector length seen (4 bytes), and a pointer to the last seen `JSArray`.
However, the vector length is limited to `BASE_CONTIGUOUS_VECTOR_LEN_MAX`, which
is `25`, so we can easily fit it in a byte, and combined with `CompactPointerTuple`,
we can fit the whole thing in 8 bytes.

* Source/JavaScriptCore/bytecode/ArrayAllocationProfile.cpp:
(JSC::ArrayAllocationProfile::updateProfile):
* Source/JavaScriptCore/bytecode/ArrayAllocationProfile.h:
(JSC::ArrayAllocationProfile::selectIndexingType):
(JSC::ArrayAllocationProfile::vectorLengthHint):
(JSC::ArrayAllocationProfile::updateLastAllocation):
(JSC::ArrayAllocationProfile::updateLastAllocationFor):
(JSC::ArrayAllocationProfile::IndexingTypeAndVectorLength::IndexingTypeAndVectorLength):
(JSC::ArrayAllocationProfile::IndexingTypeAndVectorLength::indexingType const):
(JSC::ArrayAllocationProfile::IndexingTypeAndVectorLength::vectorLength const):
(JSC::ArrayAllocationProfile::IndexingTypeAndVectorLength::operator bool const):

Canonical link: https://commits.webkit.org/260192@main
